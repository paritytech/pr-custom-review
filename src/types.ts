import github from "@actions/github"
import { RestEndpointMethodTypes } from "@octokit/plugin-rest-endpoint-methods"

export type CommitState =
  RestEndpointMethodTypes["repos"]["createCommitStatus"]["parameters"]["state"]

export type Octokit = ReturnType<typeof github.getOctokit>

export type PR = {
  number: number
  base: {
    repo: {
      name: string
      owner: {
        login: string
      }
    }
  }
  head: {
    sha: string
  }
  user: {
    login: string
  }
  html_url: string
}

export type BaseRule = {
  name: string
  condition:
    | string
    | { include: string }
    | { exclude: string }
    | { include: string; exclude: string }
  check_type: "diff" | "changed_files"
}

export type RuleCriteria = {
  min_approvals: number
  name?: string
  users?: Array<string> | null
  teams?: Array<string> | null
}

export type BasicRule = BaseRule & RuleCriteria

export type OrRule = BaseRule & {
  any: RuleCriteria[]
}

export type AndRule = BaseRule & {
  all: RuleCriteria[]
}

export type AndDistinctRule = BaseRule & {
  all_distinct: RuleCriteria[]
}

export type RuleKind = "BasicRule" | "OrRule" | "AndRule" | "AndDistinctRule"
export type Rule = BasicRule | OrRule | AndRule | AndDistinctRule

export type RulesConfigurations = {
  basic: {
    kind: "BasicRule"
    uniqueFields: ["min_approvals", "teams", "users"]
    invalidFields: ["any", "all", "all_distinct"]
  }
  and: {
    kind: "AndRule"
    uniqueFields: ["all"]
    invalidFields: ["min_approvals", "teams", "users", "any", "all_distinct"]
  }
  or: {
    kind: "OrRule"
    uniqueFields: ["any"]
    invalidFields: ["min_approvals", "teams", "users", "all", "all_distinct"]
  }
  and_distinct: {
    kind: "AndDistinctRule"
    uniqueFields: ["all_distinct"]
    invalidFields: ["min_approvals", "teams", "users", "all", "any"]
  }
}

export type Configuration = {
  rules: Rule[]
  inputs: {
    "locks-review-team": string
    "team-leads-team": string
    "action-review-team": string
  }
  prevent_review_requests:
    | {
        users: string[]
        teams: string[]
      }
    | undefined
    | null
}

export type RuleUserInfo = {
  teams: Set<string> | null
  teamsHistory?: Set<string>
}

type MatchedRuleBase = {
  name: string
  users: Map<string, RuleUserInfo>
  id: number
  kind: RuleKind
  min_approvals: number
}
export type MatchedRule =
  | (MatchedRuleBase & {
      kind: "AndRule" | "OrRule" | "BasicRule"
    })
  | (MatchedRuleBase & {
      kind: "AndDistinctRule"
      subConditions: RuleCriteria[]
    })

export class RuleSuccess {
  constructor(public rule: MatchedRule) {}
}
export class RuleFailure {
  constructor(
    public rule: MatchedRule,
    public problem: string,
    public usersToAskForReview: Map<string, RuleUserInfo>,
  ) {}
}
