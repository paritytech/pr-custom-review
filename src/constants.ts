import { CommitState } from "./github/types"
import { RulesConfigurations } from "./types"

export const configFilePath = ".github/pr-custom-review.yml"
export const actionReviewTeamFiles = [configFilePath]

export const commitStateSuccess: CommitState = "success"
export const commitStateFailure: CommitState = "failure"

export const maxGithubApiFilesPerPage = 100
export const maxGithubApiTeamMembersPerPage = 100
export const maxGithubApiReviewsPerPage = 100

export const rulesConfigurations: RulesConfigurations = {
  basic: {
    kind: "BasicRule",
    uniqueFields: ["min_approvals", "teams", "users"],
    invalidFields: ["any", "all", "all_distinct"],
  },
  and: {
    kind: "AndRule",
    uniqueFields: ["all"],
    invalidFields: ["min_approvals", "teams", "users", "any", "all_distinct"],
  },
  or: {
    kind: "OrRule",
    uniqueFields: ["any"],
    invalidFields: ["min_approvals", "teams", "users", "all", "all_distinct"],
  },
  and_distinct: {
    kind: "AndDistinctRule",
    uniqueFields: ["all_distinct"],
    invalidFields: ["min_approvals", "teams", "users", "all", "any"],
  },
}
