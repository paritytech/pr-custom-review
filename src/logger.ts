import os from "os"
import { inspect } from "util"

// GitHub by default logs N lines when the job is starting.
// N = 2 for pr-custom-review:
// 1. Run org/pr-custom-review@tag
// 2.   with:
const githubLogsInitialLineCount = 2

export interface LoggerInterface {
  doLog: (...args: any[]) => void
  log: (...args: any[]) => void
  error: (...args: any[]) => void
  warning: (...args: any[]) => void
  failure: (...args: any[]) => void
  markNextLineAsRelevantStartingLine: (delta: number) => void
}

export class Logger implements LoggerInterface {
  lineCount: number = 0
  public relevantStartingLine: number = 1

  constructor() {}

  doLog = (...args: any[]) => {
    const message = args
      .map(function (arg) {
        return typeof arg === "string" ? arg : inspect(arg)
      })
      .join(" ")
    process.stdout.write(`${message}${os.EOL}`)
    this.lineCount += (message.match(/\n/g) || "").length + 1
  }

  public log = (...args: any[]) => {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
    this.doLog(...args)
  }

  public error = (...args: any[]) => {
    // Uses escape codes for displaying the error line as red.
    // https://dustinpfister.github.io/2019/09/19/nodejs-ansi-escape-codes/
    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
    this.doLog("\n\u001b[31;1;4mERROR: ", ...args, "\u001b[39m")
  }

  public warning = (...args: any[]) => {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
    this.doLog("WARNING: ", ...args)
  }

  public failure = (...args: any[]) => {
    this.markNextLineAsRelevantStartingLine(
      // This uses delta = 2 because the error logging adds an extra newline at
      // the start
      2,
    )
    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
    this.error(...args)
  }

  public markNextLineAsRelevantStartingLine = (delta = 1) => {
    this.relevantStartingLine =
      this.lineCount + githubLogsInitialLineCount + delta
  }
}

export default Logger
