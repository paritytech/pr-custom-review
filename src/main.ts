import { getInput, setFailed } from "@actions/core"
import { context, getOctokit } from "@actions/github"

import { runChecks } from "./core"
import Logger from "./logger"
import { CommitState, PR } from "./types"

const main = function () {
  if (
    context.eventName !== "pull_request" &&
    context.eventName !== "pull_request_review"
  ) {
    setFailed(
      `Invalid event: ${context.eventName}. This action should be triggered on pull_request and pull_request_review`,
    )
    return
  }

  const logger = new Logger()

  const pr = context.payload.pull_request as PR
  const octokit = getOctokit(getInput("token", { required: true }))

  const finish = async function (state: CommitState) {
    // Fallback URL in case we are not able to detect the current job
    let detailsUrl = `${context.serverUrl}/${pr.base.repo.owner.login}/${pr.base.repo.name}/actions/runs/${context.runId}`

    if (state === "failure") {
      const jobName = process.env.GITHUB_JOB
      if (jobName === undefined) {
        logger.warning("Job name was not found in the environment")
      } else {
        // Fetch the jobs so that we'll be able to detect this step and provide a
        // more accurate logging location
        const {
          data: { jobs },
        } = await octokit.rest.actions.listJobsForWorkflowRun({
          owner: pr.base.repo.owner.login,
          repo: pr.base.repo.name,
          run_id: context.runId,
        })
        for (const job of jobs) {
          if (job.name === jobName) {
            let stepNumber: number | undefined = undefined
            const actionRepository = process.env.GITHUB_ACTION_REPOSITORY
            if (actionRepository === undefined) {
              logger.warning(
                "Action repository was not found in the environment",
              )
            } else {
              const actionRepositoryMatch = actionRepository.match(/[^/]*$/)
              if (actionRepositoryMatch === null) {
                logger.warning(
                  `Action repository name could not be extracted from ${actionRepository}`,
                )
              } else {
                const actionStep = job.steps?.find(function ({ name }) {
                  return name === actionRepositoryMatch[0]
                })
                if (actionStep === undefined) {
                  logger.warning(
                    `Failed to find ${actionRepositoryMatch[0]} in the job's steps`,
                    job.steps,
                  )
                } else {
                  stepNumber = actionStep.number
                }
              }
            }
            detailsUrl = `${job.html_url}${
              stepNumber
                ? `#step:${stepNumber}:${logger.relevantStartingLine}`
                : ""
            }`
            break
          }
        }
      }
    }

    await octokit.rest.repos.createCommitStatus({
      owner: pr.base.repo.owner.login,
      repo: pr.base.repo.name,
      sha: pr.head.sha,
      state,
      context: "Check reviews",
      target_url: detailsUrl,
      description: "Please visit Details for more information",
    })

    logger.log(`Final state: ${state}`)

    // We always exit with 0 so that there are no lingering failure statuses in
    // the pipeline for the action. The custom status created above will be the
    // one to inform the outcome of this action.
    process.exit(0)
  }

  runChecks(pr, octokit, logger)
    .then(function (state) {
      finish(state)
    })
    .catch(function (error) {
      logger.failure(error)
      finish("failure")
    })
}

main()
