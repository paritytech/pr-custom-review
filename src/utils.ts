import assert from "assert";

export const envVar = (name: string) => {
  const val = process.env[name];
  if (typeof val !== "string") {
    throw new Error(`${name} was not found in the environment variables`);
  }
  return val;
};

export const envNumberVar = (name: string) => {
  const val = process.env[name];
  assert(val, `${name} was not found in the environment variables`);
  const valNumber = parseInt(val);
  assert(!isNaN(valNumber), `${name} is not a number`);
  return valNumber;
};

export const delayMilliseconds = (duration: number) => new Promise<void>((resolve) => setTimeout(resolve, duration));

export class Ok<T> {
  constructor(public value: T) {}
}
export class Err<T> {
  constructor(public value: T) {}
}

/* eslint-disable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call */
const normalizers = {
  symbol: (value: any) => value.toString(),
  bigint: (value: any) => value.toString(),
  undefined: () => undefined,
  function: () => undefined,
  boolean: (value: any) => value,
  number: (value: any) => value,
  string: (value: any) => value,
  object: (value: any, previousObjects: unknown[] = []) => {
    if (value === null) {
      return;
    }

    previousObjects = previousObjects.concat([value]);

    const isArray = Array.isArray(value);
    const isIterable = !isArray && Symbol.iterator in value;
    const objAsArray = isArray ? value : isIterable ? Array.from(value as Iterable<unknown>) : undefined;

    if (objAsArray === undefined && !(value instanceof Error)) {
      const asString =
        typeof value.toString === "function" && value.toString.length === 0 ? value.toString() : undefined;
      if (typeof asString === "string" && asString !== "[object Object]") {
        return asString;
      }
    }

    const { container, output } = (() => {
      if (isIterable) {
        const iteratorContainer = { type: value.constructor.name, items: [] };
        return { container: iteratorContainer, output: iteratorContainer.items };
      }
      const outputObj = objAsArray === undefined ? {} : [];
      return { container: outputObj, output: outputObj };
    })();

    const sourceObj = objAsArray ?? value;
    for (const key of Object.getOwnPropertyNames(sourceObj)) {
      setNormalizedKeyValue(sourceObj, output, key, previousObjects);
    }

    if (Object.keys(output).length > 0) {
      return container;
    }
  },
};

const setNormalizedKeyValue = (source: any, output: any, key: any, previousObjects: unknown[]) => {
  if (previousObjects.indexOf(source[key]) !== -1) {
    return "[Circular]";
  }

  const value = normalizeValue(source[key], previousObjects);
  if (value === undefined) {
    return;
  }

  output[key] = value;
};

export const normalizeValue = (value: unknown, previousObjects: unknown[] = []): unknown =>
  normalizers[typeof value](value, previousObjects);
