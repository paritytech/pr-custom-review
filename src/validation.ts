import Joi from "joi";

import { AndDistinctRule, AndRule, BasicRule, Configuration, OrRule, RuleCriteria } from "./types";

const ruleCriterion = ({
  isMinApprovalsAllowed,
  isNameOptional,
}: {
  isMinApprovalsAllowed: boolean;
  isNameOptional: boolean;
}) => {
  let name = Joi.string();
  if (isNameOptional) {
    name = name.optional().allow(null);
  } else {
    name = name.required();
  }

  return {
    name,
    users: Joi.array().items(Joi.string()).optional().allow(null),
    teams: Joi.array().items(Joi.string()).optional().allow(null),
    ...(isMinApprovalsAllowed ? { min_approvals: Joi.number().min(1).required() } : {}),
  };
};

const ruleCriterionArraySchema = Joi.array()
  .items(Joi.object<RuleCriteria>().keys(ruleCriterion({ isMinApprovalsAllowed: true, isNameOptional: true })))
  .required();

const includeConditionSchema = Joi.string().required();
const excludeConditionSchema = Joi.string().required();
const commonRuleSchema = {
  name: Joi.string().required(),
  condition: Joi.alternatives([
    includeConditionSchema,
    Joi.object().keys({ include: includeConditionSchema }),
    Joi.object().keys({ exclude: excludeConditionSchema }),
    Joi.object().keys({ include: includeConditionSchema, exclude: excludeConditionSchema }),
  ]).required(),
  check_type: Joi.string().valid("diff", "changed_files").required(),
};

const ruleSchema = Joi.alternatives([
  Joi.object<BasicRule>().keys({
    ...commonRuleSchema,
    ...ruleCriterion({ isMinApprovalsAllowed: true, isNameOptional: false }),
  }),
  Joi.object<AndRule>().keys({ ...commonRuleSchema, all: ruleCriterionArraySchema }),
  Joi.object<OrRule>().keys({ ...commonRuleSchema, any: ruleCriterionArraySchema }),
  Joi.object<AndDistinctRule>().keys({ ...commonRuleSchema, all_distinct: ruleCriterionArraySchema }),
]);

export const configurationSchema = Joi.object<Configuration>().keys({
  "locks-review-team": Joi.string().required(),
  "team-leads-team": Joi.string().required(),
  "action-review-team": Joi.string().required(),
  rules: Joi.array().items(ruleSchema).required(),
  "prevent-review-request": Joi.object<Configuration["prevent-review-request"]>()
    .keys({
      users: Joi.array().items(Joi.string()).optional().allow(null),
      teams: Joi.array().items(Joi.string()).optional().allow(null),
    })
    .optional()
    .allow(null),
});
