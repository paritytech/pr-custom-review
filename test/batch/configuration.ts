import { Octokit } from "@octokit/rest"
import nock from "nock"
import YAML from "yaml"

import {
  basePR,
  changedFilesApiPath,
  condition,
  configFileContentsApiPath,
  githubApi,
  defaultTeamsNames,
  prApiPath,
  rulesExamples,
} from "test/constants"
import Logger from "test/logger"

import { rulesConfigurations } from "src/constants"
import { runChecks } from "src/core"

const setup = function ({ rules }: { rules?: Array<Record<string, unknown>> }) {
  nock(githubApi)
    .get(configFileContentsApiPath)
    .reply(200, {
      content: Buffer.from(
        YAML.stringify({ ...defaultTeamsNames, rules }),
      ).toString("base64"),
    })
}

describe("Configuration", function () {
  let logger: Logger
  let octokit: Octokit
  let logHistory: string[]

  beforeEach(function () {
    nock.disableNetConnect()
    logHistory = []
    logger = new Logger(logHistory)
    octokit = new Octokit()
    nock(githubApi)
      .get(prApiPath)
      .matchHeader("accept", "application/vnd.github.v3.diff")
      .reply(200, condition)
    nock(githubApi)
      .get(changedFilesApiPath)
      .reply(200, [{ filename: condition }])
  })

  for (const { kind, invalidFields } of Object.values(rulesConfigurations)) {
    const goodRule = rulesExamples[kind]

    for (const invalidField of invalidFields) {
      const invalidFieldValidValue = (function () {
        switch (invalidField) {
          case "all":
          case "all_distinct":
          case "teams":
          case "users":
          case "any": {
            return []
          }
          case "min_approvals": {
            return 1
          }
          default: {
            const exhaustivenessCheck: never = invalidField
            throw new Error(
              `invalidField is not handled: ${exhaustivenessCheck}`,
            )
          }
        }
      })()

      it(`Rule kind ${kind} does not allow invalid field ${invalidField}`, async function () {
        const badRule = { ...goodRule, [invalidField]: invalidFieldValidValue }

        setup({ rules: [badRule] })

        expect(await runChecks(basePR, octokit, logger)).toBe("failure")

        expect(logHistory).toMatchSnapshot()
      })
    }
  }

  for (const [kind, exampleRule] of Object.entries(rulesExamples)) {
    for (const [invalidValue, description] of [
      [0, "less than 1"],
      [null, "null"],
    ]) {
      it(`min_approvals is invalid for ${kind} if it is ${description}`, async function () {
        setup({
          rules: [
            {
              ...exampleRule,
              ...("min_approvals" in exampleRule
                ? { min_approvals: invalidValue }
                : "all" in exampleRule
                ? { all: [{ min_approvals: invalidValue }] }
                : "any" in exampleRule
                ? { any: [{ min_approvals: invalidValue }] }
                : "all_distinct" in exampleRule
                ? { all_distinct: [{ min_approvals: invalidValue }] }
                : {}),
            },
          ],
        })

        expect(await runChecks(basePR, octokit, logger)).toBe("failure")

        expect(logHistory).toMatchSnapshot()
      })
    }
  }

  afterEach(function () {
    nock.cleanAll()
    nock.enableNetConnect()
  })
})
