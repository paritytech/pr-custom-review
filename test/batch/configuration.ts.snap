// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configuration Rule kind AndDistinctRule does not allow invalid field all 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Rule kind AndDistinctRule does not allow invalid field any 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Rule kind AndDistinctRule does not allow invalid field min_approvals 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Rule kind AndDistinctRule does not allow invalid field teams 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Rule kind AndDistinctRule does not allow invalid field users 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Rule kind AndRule does not allow invalid field all_distinct 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Rule kind AndRule does not allow invalid field any 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Rule kind AndRule does not allow invalid field min_approvals 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Rule kind AndRule does not allow invalid field teams 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Rule kind AndRule does not allow invalid field users 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Rule kind BasicRule does not allow invalid field all 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Rule kind BasicRule does not allow invalid field all_distinct 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Rule kind BasicRule does not allow invalid field any 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Rule kind OrRule does not allow invalid field all 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Rule kind OrRule does not allow invalid field all_distinct 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Rule kind OrRule does not allow invalid field min_approvals 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Rule kind OrRule does not allow invalid field teams 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Rule kind OrRule does not allow invalid field users 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration min_approvals is invalid for AndDistinctRule if it is less than 1 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration min_approvals is invalid for AndDistinctRule if it is null 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration min_approvals is invalid for AndRule if it is less than 1 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration min_approvals is invalid for AndRule if it is null 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration min_approvals is invalid for BasicRule if it is less than 1 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration min_approvals is invalid for BasicRule if it is null 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration min_approvals is invalid for OrRule if it is less than 1 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration min_approvals is invalid for OrRule if it is null 1`] = `
Array [
  "Running checks for repo#1",
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0]\\" does not match any of the allowed types] {
  _original: {
    'locks-review-team': 'team',
    'team-leads-team': 'team2',
    'action-review-team': 'team3',
    rules: [ [Object] ]
  },
  details: [
    {
      message: '\\"rules[0]\\" does not match any of the allowed types',
      path: [Array],
      type: 'alternatives.match',
      context: [Object]
    }
  ]
}",
]
`;
